### PR TITLE
MEP_oM: Object migration

### LINKED ISSUE #952: MEP_oM: Initial Object Creation
#### Description:

Initial MEP_oM contributions will consist of the following elements and interfaces. 

Creation of MEP objects:
- [ ] Duct
- [ ] Pipe
- [ ] Wire
- [ ] Mechanical System
- [ ] System
- [ ] Node
- [ ] SystemType

Creation of MEP Interfaces
- [ ] IFlow

Enums
- [ ] FluidType

Material Fragments + Material Interfaces
- [ ] Duct Material
- [ ] Pipe Material
- [ ] Wire Material
- [ ] Insulation Material
- [ ] Lining Material
- [ ] IInsulationMaterial
- [ ] IMEPMaterial

Section Properties + Section Interfaces
- [ ] DuctSectionProperty
- [ ] PipeSectionProperty
- [ ] WireSectionProperty
- [ ] SectionProfile
- [ ] IFlowSectionProperty

MEP_Engine connecting issue: https://github.com/BHoM/BHoM_Engine/issues/1908

### PR DESCRIPTION
**Pull request by michaelhoehn**
### NOTE: Depends on 

https://github.com/BHoM/BHoM_Engine/pull/1911

   
### Issues addressed by this PR

Closes #952 

### Test files

https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM/MEP_oM/200728_MEP_oM%20Testing.gh?csf=1&web=1&e=1rmXkx

### CODE REVIEW

**Comment by michaelhoehn** [comment]
> Section profiles code to be updated to include shapeprofile, liningthickness and insulationthickness:

Thanks @kayleighhoude, the PR dependency has been updated to include the matching MEP_Engine PR. 

This functionality exists within the MEP_Engine as it's creating the composite sections. Without it, the example script will replace with the explicit SectionProfile that hosts the profiles themselves. 

If anyone else has this same error, please make sure to build both oM + Engine branches to get the full experience 😄

File: MEP_oM/SurfaceProperties/ConstantThickness.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+using BH.oM.Base;
+using BH.oM.MEP.MaterialFragments;
+using System.ComponentModel;
+using BH.oM.Quantities.Attributes;
+
+namespace BH.oM.MEP.SurfaceProperties
+{
+    [Description("Simplest property for 2D analytical elements with a constant thickness and homogenous material across the whole element.")]
+    public class ConstantThickness : BHoMObject, ISurfaceProperty
```
Comment (IsakNaslundBh) [inline MEP_oM/SurfaceProperties/ConstantThickness.cs:31]: Is this class needed for MEP? Would assume the env panels to do the job needed for the BEnv/MEP side of things, and that, as far as I know, is making use of the physical construction. Would not add in this class unless this is targeting something completely different that I am not aware of.

@FraserGreenroyd

File: MEP_oM/SurfaceProperties/ISurfaceProperty.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+using BH.oM.Base;
+using BH.oM.MEP.MaterialFragments;
+using System.ComponentModel;
+
+namespace BH.oM.MEP.SurfaceProperties
+{
+    [Description("Base interface for properties for 2D finite element structural objects such as Panels and FEMeshes.")]
+    public interface ISurfaceProperty : IBHoMObject
```
Comment (IsakNaslundBh) [inline MEP_oM/SurfaceProperties/ISurfaceProperty.cs:30]: See comment above regarding the constant thickness.

File: MEP_oM/SectionProperties/SectionProfile.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.ComponentModel;
+
+using BH.oM.Base;
+using BH.oM.Geometry.ShapeProfiles;
+
+namespace BH.oM.MEP.SectionProperties
+{
+    public class SectionProfile : BHoMObject, IBHoMObject
```
Comment (IsakNaslundBh) [inline MEP_oM/SectionProperties/SectionProfile.cs:35]: Adding this is not technically incorrect in any shape or form, but wonder if it is simpler to just put all of these three profiles as separate properties on the duct, pile and wire properties as needed. 
Ofc, good to keep things organised, but wonder if this additional layer actually helps. It will mean one more layer you have to explode/get properties from when you want to extract the dimensions of the pipe/duct etc. Also, in terms of naming of this, this sounds very struyctures related. If MEP are using the same name (Section) as structures do, that is ofc fine, but want to make sure that is the case @kayleighhoude

File: MEP_oM/SectionProperties/PipeSectionProperty.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    public class PipeSectionProperty : BHoMObject, IFlowSectionProperty, IImmutable
+    {
+        [Description("The pipe material is the primary material that the pipe is composed of (copper, black steel, PEX)")]
+        public IMEPMaterial PipeMaterial { get; set; }
+
+        [Description("The insulation is the layer of material outside of the pipe material, meant to insulate the internal conditions of the material being conveyed.")]
+        public IMEPMaterial InsulationMaterial { get; set; }
+
+        [Description("The section profile of the object that will determine its use within a System.")]
+        public virtual SectionProfile SectionProfile { get; set; }
```
Comment (IsakNaslundBh) [inline MEP_oM/SectionProperties/PipeSectionProperty.cs:44]: should be get only to make sure things such as areas can not be invalidated.

File: MEP_oM/SectionProperties/SectionProfile.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        [Description("ShapeProfile of the element itself. This is the base ShapeProfile that determines the overall shape of the IFlow object.")]
+        public virtual IProfile ElementProfile { get; set; }
```
Comment (IsakNaslundBh) [inline MEP_oM/SectionProperties/SectionProfile.cs:38]: If kept, this needs to be made immutable, and constructor needed to be added and all the properties need to be made get only.

File: MEP_oM/SectionProperties/SectionProfile.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+        [Description("ShapeProfile of the element's interior lining.")]
+        public virtual IProfile LiningProfile { get; set; }
```
Comment (IsakNaslundBh) [inline MEP_oM/SectionProperties/SectionProfile.cs:41]: as above

File: MEP_oM/SectionProperties/SectionProfile.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+        [Description("ShapeProfile of the element's exterior insulation.")]
+        public virtual IProfile InsulationProfile { get; set; }
```
Comment (IsakNaslundBh) [inline MEP_oM/SectionProperties/SectionProfile.cs:44]: as above

File: MEP_oM/SectionProperties/DuctSectionProperty.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        public virtual IMEPMaterial LiningMaterial { get; set; }
+
+        [Description("Hydraulic Diameter allows you to calculate the round equivalent hydraulic diameter for a non-round duct (rectangular/square).")]
+        public virtual double HydraulicDiameter { get; }
+
+        [Description("CircularEquivalentDiameter is the equivalent size for a duct if it were to be round (for size/capacity comparison).")]
+        public virtual double CircularEquivalentDiameter { get; }
+
+        [Description("The section profile of the object that will determine its use within a System.")]
+        public virtual SectionProfile SectionProfile { get; set; }
```
Comment (IsakNaslundBh) [inline MEP_oM/SectionProperties/DuctSectionProperty.cs:53]: should be get only to make sure things such as areas can not be invalidated.

File: MEP_oM/Elements/Wire.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+using BH.oM.Base;
+using BH.oM.Analytical.Elements;
+using BH.oM.MEP.SectionProperties;
+using BH.oM.Dimensional;
+
+
+namespace BH.oM.MEP.Elements
+{
+    [Description("Wire object to work within an MEP systems.")]
+    public class Wire : BHoMObject, IFlow, ILink<Node>, IElement1D, IElementM
```
Comment (IsakNaslundBh) [inline MEP_oM/Elements/Wire.cs:40]: After comments and discussions about this before about wires not necessarily being straight segments, worth considering if this should be a 2-noded element, implementing `ILink<Node>` or instead of having a start and end node, it could store a `ICurve`. Start and end nodes, or at least points, could be extracted from this curve as query method for the wire.

Do not know enough of this to say what is right and wrong, but putting as a general comment.

File: MEP_oM/SurfaceProperties/ConstantThickness.cs (hunk 1 above)
Comment (FraserGreenroyd) [reply MEP_oM/SurfaceProperties/ConstantThickness.cs:31]: I would agree, though only our panels make use of physical constructions, so we might just need to see where this is being used and re-evaluate as appropriate?

Comment (IsakNaslundBh) [reply MEP_oM/SurfaceProperties/ConstantThickness.cs:31]: Searched for all references, and could not find any in BHoM at all. If that is just due to the fact that some object has not yet been added that will make use of it, I would ofc not have anything against this being added.

The comment was more a check if it was added to mirror the structures suit of properties with no use yet in MEP/BEnv.

File: MEP_oM/Elements/Wire.cs (hunk 9 above)
Comment (IsakNaslundBh) [reply MEP_oM/Elements/Wire.cs:40]: Could be similar to structures where we have the `IFramingElement`s being the phsyical representation, able to be any curve shape, and the `Bar`s which are the idealised representations, being straight segements between two nodes. We have a method in structures to go from an `IFramingElement` to a list of `Bar`s, that for straight, prismatic elements only give a single bar back, but for arc shaped ones etc gives back multiple.

Fundamentally, might need to states of representations for different cases.

File: MEP_oM/SectionProperties/SectionProfile.cs (hunk 3 above)
Comment (michaelhoehn) [reply MEP_oM/SectionProperties/SectionProfile.cs:35]: The class name seems logical to me but would happily defer to an MEP expert on this. I personally like how they align across disciplines, but happy to change if confusing. As far as the necessity for the class itself, I completely understand the suggestion to expose these properties one level up. Happy to refactor the way the class is working upon further testing but would like to see how it's being used prior to making another major change to the class structure if that's alright 😄

File: MEP_oM/Elements/System.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("A collection of FlowObjects to compile within a specified systemType.")]
+        public virtual List<IFlow> Objects { get; set; } = null;
+
+        [Description("A specific system type to be applied to the flow objects. The system type describes the material conveyed by the system (supply air, cold water), therefore determining the behaviours of the system")]
+        public virtual SystemType SystemType { get; set; } = null;
```
[the remark this answers is not in the document]
Comment (michaelhoehn) [reply MEP_oM/Elements/System.cs:47]: Will be treated as ISystemType as an empty interface for now per your other comments.

File: MEP_oM/Elements/FlowSystem.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+{
+    [Description("A collection of IFlow Objects to be utilised throughout a defined system.")]
+    public class FlowSystem : BHoMObject, IElement1D, IElementM
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("A collection of FlowObjects to compile within a specified systemType.")]
+        public virtual List<IFlow> FlowObjects { get; set; } = null;
```
Comment (pawelbaran) [inline MEP_oM/Elements/FlowSystem.cs:44]: Just curious: will `IFlow` interface be implemented by the equipment? Or will the presence of non-linear objects in the system be handled differently?

File: MEP_oM/Elements/IFlow.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+namespace BH.oM.MEP.Elements
+{
+    [Description("Base interface for all flow-based objects. These objects are capable of containing a material or element that flows through the object.")]
+    public interface IFlow : IBHoMObject, IElement1D, IElementM
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("The point at which the Duct object begins.")]
```
Comment (pawelbaran) [inline MEP_oM/Elements/IFlow.cs:43]: It is not a duct

File: MEP_oM/Elements/IFlow.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        Node StartNode { get; set; }
+
+        [Description("The point at which the Duct object ends.")]
```
Comment (pawelbaran) [inline MEP_oM/Elements/IFlow.cs:46]: It is not a duct

File: MEP_oM/Elements/MechanicalSystem.cs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+namespace BH.oM.MEP.Elements
+{
+    [Description("Mechanical systems are qualified by their ability to convey air, water, etc to a building/room/area")]
+    public class MechanicalSystem : BHoMObject, ISystemType
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        //Temporary properties require additional user input.
```
Comment (pawelbaran) [inline MEP_oM/Elements/MechanicalSystem.cs:44]: Why is that here?

File: MEP_oM/Elements/Node.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+using BH.oM.Base;
+using BH.oM.Geometry;
+using BH.oM.Analytical.Elements;
+using BH.oM.Dimensional;
+
+namespace BH.oM.MEP.Elements
+{
+    public class Node : BHoMObject, IElement0D, INode
```
Comment (pawelbaran) [inline MEP_oM/Elements/Node.cs:30]: Descriptions are missing in this file

File: MEP_oM/Elements/Wire.cs (hunk 9 above)
Comment (pawelbaran) [reply MEP_oM/Elements/Wire.cs:40]: I would agree with @IsakNaslundBh, would be nice to have an analytical representation of MEP objects (done in this PR) and a seperate physical one - I would expect this issue coming back like a boomerang, e.g. in case of fittings (do we always need to know their exact geometry?) etc.

File: MEP_oM/MaterialFragments/DuctMaterial.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using BH.oM.Base;
+
+namespace BH.oM.MEP.MaterialFragments
+{
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/DuctMaterial.cs:33]: header and /***************************************************/ missing

File: MEP_oM/MaterialFragments/DuctMaterial.cs [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+    public class DuctMaterial : BHoMObject, IMEPMaterial
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/DuctMaterial.cs:34]: Description missing

File: MEP_oM/MaterialFragments/InsulationMaterial.cs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using BH.oM.Base;
+
+namespace BH.oM.MEP.MaterialFragments
+{
+    public class InsulationMaterial : BHoMObject, IMEPMaterial, IInsulationMaterial
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/InsulationMaterial.cs:34]: Description missing

File: MEP_oM/MaterialFragments/InsulationMaterial.cs [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        [Description("RValue is the measure of the resistance of conductive heat loss by the insulation material.")]
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/InsulationMaterial.cs:36]: Header and /***************************************************/ missing

File: MEP_oM/MaterialFragments/LiningMaterial.cs [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using BH.oM.Base;
+
+namespace BH.oM.MEP.MaterialFragments
+{
+    public class LiningMaterial : BHoMObject, IMEPMaterial, IInsulationMaterial
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/LiningMaterial.cs:34]: Description missing

File: MEP_oM/MaterialFragments/LiningMaterial.cs [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        [Description("RValue is the measure of the resistance of conductive heat loss by the lining material.")]
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/LiningMaterial.cs:36]: Header and /***************************************************/ missing

File: MEP_oM/MaterialFragments/PipeMaterial.cs [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using BH.oM.Base;
+
+namespace BH.oM.MEP.MaterialFragments
+{
+    public class PipeMaterial : BHoMObject, IMEPMaterial
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/PipeMaterial.cs:34]: Description missing

File: MEP_oM/MaterialFragments/PipeMaterial.cs [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        [Description("Roughness is a measure of the irregularities on the surface of a pipe.")]
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/PipeMaterial.cs:36]: Header and /***************************************************/ missing

File: MEP_oM/MaterialFragments/WireMaterial.cs [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using BH.oM.Base;
+
+namespace BH.oM.MEP.MaterialFragments
+{
+    public class WireMaterial : BHoMObject, IMEPMaterial
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/WireMaterial.cs:34]: Description missing

File: MEP_oM/MaterialFragments/WireMaterial.cs [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        [Description("The degree to which a specific material conducts electricity, calculated as the ratio of the current density and the material to the electric field that causes the flow of current.")]
```
Comment (pawelbaran) [inline MEP_oM/MaterialFragments/WireMaterial.cs:36]: Header and /***************************************************/ missing

File: MEP_oM/SectionProperties/DuctSectionProperty.cs [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.ComponentModel;
+
+using BH.oM.Base;
+using BH.oM.MEP.MaterialFragments;
+
+namespace BH.oM.MEP.SectionProperties
+{
+    public class DuctSectionProperty : BHoMObject, IFlowSectionProperty, IImmutable
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/DuctSectionProperty.cs:35]: Description missing

File: MEP_oM/SectionProperties/DuctSectionProperty.cs [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        [Description("The duct material is the primary material that the duct is composed of (galvanized sheet metal, aluminium)")]
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/DuctSectionProperty.cs:37]: Header and /***************************************************/ missing

File: MEP_oM/SectionProperties/IFlowSectionProperty.cs [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Base;
+
+namespace BH.oM.MEP.SectionProperties
+{
+    public interface IFlowSectionProperty : IBHoMObject
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/IFlowSectionProperty.cs:27]: Description missing (not sure if needed for interfaces @FraserGreenroyd?)

File: MEP_oM/SectionProperties/PipeSectionProperty.cs [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.ComponentModel;
+
+using BH.oM.Base;
+using BH.oM.MEP.MaterialFragments;
+
+namespace BH.oM.MEP.SectionProperties
+{
+    public class PipeSectionProperty : BHoMObject, IFlowSectionProperty, IImmutable
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/PipeSectionProperty.cs:35]: Description missing

File: MEP_oM/SectionProperties/PipeSectionProperty.cs [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        [Description("The pipe material is the primary material that the pipe is composed of (copper, black steel, PEX)")]
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/PipeSectionProperty.cs:37]: Header and /***************************************************/ missing

File: MEP_oM/SectionProperties/PipeSectionProperty.cs [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+        public IMEPMaterial PipeMaterial { get; set; }
+
+        [Description("The insulation is the layer of material outside of the pipe material, meant to insulate the internal conditions of the material being conveyed.")]
+        public IMEPMaterial InsulationMaterial { get; set; }
+
+        [Description("The section profile of the object that will determine its use within a System.")]
+        public virtual SectionProfile SectionProfile { get; }
+
+        [Description("This area takes the element's thickness into account to determine the actual area of the 'solid' portion of the ShapeProfile.")]
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/PipeSectionProperty.cs:46]: `SectionProfile` or `ShapeProfile`? (same to other section properties)

File: MEP_oM/SectionProperties/SectionProfile.cs [hunk 32]
```diff
@@ -0,0 +1,46 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.ComponentModel;
+
+using BH.oM.Base;
+using BH.oM.Geometry.ShapeProfiles;
+
+namespace BH.oM.MEP.SectionProperties
+{
+    public class SectionProfile : BHoMObject, IBHoMObject, IImmutable
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/SectionProfile.cs:35]: Description missing

File: MEP_oM/SectionProperties/SectionProfile.cs [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        [Description("ShapeProfile of the element itself. This is the base ShapeProfile that determines the overall shape of the IFlow object.")]
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/SectionProfile.cs:37]: Header and /***************************************************/ missing

File: MEP_oM/SectionProperties/WireSectionProperty.cs [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.ComponentModel;
+
+using BH.oM.Base;
+using BH.oM.MEP.MaterialFragments;
+
+namespace BH.oM.MEP.SectionProperties
+{
+    public class WireSectionProperty : BHoMObject, IFlowSectionProperty, IImmutable
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/WireSectionProperty.cs:35]: Description missing

File: MEP_oM/SectionProperties/WireSectionProperty.cs [hunk 35]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        [Description("The wire material is the primary material that the wire is composed of (copper, aluminium).")]
```
Comment (pawelbaran) [inline MEP_oM/SectionProperties/WireSectionProperty.cs:37]: Header and /***************************************************/ missing

**Comment by pawelbaran** [comment]
By the way, aren't we planning a cable tray object?

**Comment by kayleighhoude** [comment]
> By the way, aren't we planning a cable tray object?

cable tray and conduit are next up because they convey wires (we considered that to be the more core MEP system object)

File: MEP_oM/Elements/FlowSystem.cs (hunk 11 above)
Comment (FraserGreenroyd) [reply MEP_oM/Elements/FlowSystem.cs:44]: Not by the equipment at this stage, but the equipment is made up of parts and some of the parts may implement `IFlow` as appropriate.

File: MEP_oM/SectionProperties/IFlowSectionProperty.cs (hunk 28 above)
Comment (FraserGreenroyd) [reply MEP_oM/SectionProperties/IFlowSectionProperty.cs:27]: Not needed as should be implemented by each object. Unless a description is generic across all implementing objects :smile:

**Review by FraserGreenroyd — APPROVED** [review]
I am now happy with these object definitions and am happy to merge this to allow further build out of the MEP space :smile: great work @michaelhoehn
